Add CI workflow to compute diff between files dist files

Kocal opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Issues        | Fix #... <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT

## 🚨 Before merging, we should drop the commit that modify files (in order to generate the diff table)

<!--
Replace this notice by a description of your feature/bugfix.
This will help reviewers and should be a good start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - For new features, provide some code snippets to help understand usage.
 - Features and deprecations must be submitted against branch main.
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
 - Never break backward compatibility (see https://symfony.com/bc).
-->

This PR is purely internal, and aims to display the `assets/dist/*.{js,css}` files diff between `2.x` and a pull-request. Similar to https://github.com/marketplace/actions/pkg-size-action, but fully internal.

I wanted a tool that display dist files size diff for each pull-request, because I was a bit afraid of changes done in https://github.com/symfony/ux/pull/2160.
 
When a PR is opened, it check dist files between the base branch (`2.x`) and the pull request, and it create a GitHub comment. The comment is created by https://github.com/marocchino/sticky-pull-request-comment, and is automatically updated depending of the check state.
If any diff between dist files, then a table is displayed, with a line per file. It shows the  original size and compressed (gzip and brotli) sizes, and also a difference in %. 

## States

_Currently not working on this repository, but you can see them on https://github.com/Kocal/symfony-ux/pull/1_ 

### 
<img width="931" alt="Capture d’écran 2024-10-13 à 11 45 10" src="https://github.com/user-attachments/assets/288523cc-f8fa-48a7-a1e1-08174403b54d">
When opening a PR

### When an issue happened
<img width="914" alt="Capture d’écran 2024-10-13 à 11 45 19" src="https://github.com/user-attachments/assets/ae0dae2a-1998-498c-935f-e5d42f78889e">

### When there is no difference between base and PR
<img width="949" alt="Capture d’écran 2024-10-13 à 11 25 31" src="https://github.com/user-attachments/assets/0a7a56d8-1183-4341-99f6-ca386f552298">

### When there is difference between base and PR
<img width="913" alt="image" src="https://github.com/user-attachments/assets/9a5ccaf0-96cf-4189-9bc3-2e35157db70a">

Kocal committed:
Add CI workflow to compute diff between files dist files

Kocal commented:
It looks like there are permissions issues, even after added `permissions.pull-requests: write` like [mentioned in the documentation](https://github.com/marocchino/sticky-pull-request-comment#basic):
<img width="1521" alt="image" src="https://github.com/user-attachments/assets/fc5d4b1a-192c-4287-836f-e2116dc3444b">

Even if I use `permissions: 'write-all'` it does not work either:
<img width="374" alt="image" src="https://github.com/user-attachments/assets/5b614fe5-2e10-471c-9c0e-4c8f814eeeb0">

I suspect a configuration at organization-level...? 

@kbond do you think you can do something about that please? 🙏🏻 

/cc @nicolas-grekas or maybe you Nicolas since you setup the same action on Symfony recipes repositories :) 

Thanks!

javiereguiluz commented on .github/generate-dist-files-size-diff.mjs:
```diff
@@ -0,0 +1,118 @@
+    output += 'Thanks for the PR! Here is the difference in size of the dist files between the base and the PR.\n';
+    output += 'Please review the changes and make sure they are expected.\n\n';
+    output += `<table>
+    <thead><tr><th>File</th><th>Diff (B)</th><th>Diff (%)</th></tr></thead>
```
I would simplify the output to display something more similar to pck-size action, which I think it's very readable. Something like this:

```js
    <thead><tr><th>File</th><th>Before (Size / Gzip)</th><th>After (Size / Gzip)</th></tr></thead>
    <tbody>`;
    for (const [file, details] of files.entries()) {
        output += `<tr>
        <td><code>${file}</code> ${details.state === 'added' ? '(new)' : (details.state === 'removed' ? '(deleted)' : '')}</td>
        <td>
            <code>${formatBytes(base[file]?.size || 0)}</code> / <code>${formatBytes(base[file]?.size_gz || 0)}</code>
        </td>
        <td align="right">
            <sup>${formatDiffPercent(details.diff_percent_size)}</sup> (<code>${formatBytes(pr[file]?.size || 0)}</code>)
            / <sup>${formatDiffPercent(details.diff_percent_size_gz)}</sup> <code>${formatBytes(pr[file]?.size_gz || 0)}</code>)
        </td>
    </tr>
```

Kocal replied:
This is what I've did first, but that was ugly because our filenames are very very long (but I think I can do something about it) and it was breaking the sizes into multiple lines.

I will rework that part, thanks

Kocal replied:
I re-used your code and this is what happens:
<img width="979" alt="image" src="https://github.com/user-attachments/assets/0339ecba-b79a-4d1e-ae92-7673349e37af">

As you can see, because of line breaks the display is not very nice :/

Kocal replied:
Also, it does not even contain the brotli-compressed size of dist files, which I really want to keep (since it's more efficient than gzip and zstd)

Kocal replied:
I reworked the table rendering to display info in a smarter way:
<img width="878" alt="image" src="https://github.com/user-attachments/assets/a86158a1-fa4f-48c7-8ed9-795cdfa4ff1e">


You don't have useless data like `0 B` or `-+100%` displayed anymore, instead it shows you if the file has been added or removed (without displaying size) and the delta (%) only if the file has been changed.
I've regrouped files by packages, and each packages and files has a link.

kbond commented:
Shouldn't this job only run if a dist file has been changed?

Kocal commented:
> Shouldn't this job only run if a dist file has been changed?

Fixed by using the following config:
```yaml
on:
  pull_request:
    paths:
      - 'src/*/assets/dist/**'
      - 'src/*/src/Bridge/*/assets/dist/**'

```

I will push soon, thanks

Kocal commented:
I've changed the table rendering after @javiereguiluz's comment, it now looks like this:
<img width="878" alt="image" src="https://github.com/user-attachments/assets/671486e6-a4b2-4290-b5e5-9de0445e5de6">

We still need to find out why there is this `Resource not accessible by integration` error, and we are good IMO

jmsche commented on .github/generate-dist-files-size-diff.mjs:
```diff
@@ -0,0 +1,182 @@
+        return 0;
+    }
+
+    return Number(((from - to) / to * -100).toFixed(1));
```
I think you should divide by `from`, not `to`, to compare with the previous file size?

Kocal replied:
I don't think so:
- if the file was previously `1000` bytes and is now `1500` bytes, I want to display an increase of `33,33%`
- in the opposite, if the file was previously `1500` bytes and is now `1000` bytes, I want to display a decrease of `50%`
<img width="326" alt="image" src="https://github.com/user-attachments/assets/229f8710-902c-4a88-94b8-b8c7c39c7102">

Or did I miss something?

jmsche replied:
To me it's the other way:

- previously it was 1000 bytes, now 1500, so it's +50% compared to the previous value
- and if it was 1500 bytes and now 1000, it's -33,33% compared to the previous value

javiereguiluz replied:
@jmsche is right. Going from 1,000 to 1,500 is a +50%, not a +33% _(you are adding 500 to 1,000 ... so you are adding 50% of the original value)_

Kocal replied:
... 🤦🏻 

Thanks guys 😅

javiereguiluz commented:
@Kocal I like a lot what you are doing here. Thanks.

Some additional comments ... but I could be wrong, so feel free to ignore them:

* I'd remove Brotli because it's just "the same as Gzip but a bit better", so it doesn't add any relevant information
* I'd round percentages and remove decimals to make them easier to understand: `-4%` is better to me than `-3.9%` and `+74%` is better than `+74.3%`
* I'd put the percentages first, because it's the most relevant info by far. E.g. I don't care that the new file size is `5.56 KB`, I just want to quickly check that it's a `-27%` file size change.

Thanks!

Kocal commented:
Thanks Javier, the more we speak about Brotli and the more I think we can remove it (btw [pkg-size-action](https://github.com/marketplace/actions/pkg-size-action) can display Brotli sizes).  Gzip is the most common compression standard for the moment, while Brotli is — I believe —  still a "niche" thing. Brotli users know that their files will be smaller than gzipped files.

I agree for percentage rounding.

When it comes to putting percentages first, I don't really agree with you. Yes, percentage differences are important, but when the browser goes to download a resource, it's the size of the resource that matters, not the percentage difference.
If a file is reduced from 1 MB to 500 KB, we gain 50%, but more importantly we save 500 KB (and that's a really huge gain). If a file is reduced from 10 Kb to 5 Kb, you also gain 50%, but you only save 5KB (which still has an impact, but a much smaller one).
But then, your point of view is valid too and I can understand it, I think here it's two different ways of thinking :)

Kocal commented:
Thanks for your reviews, the diff table now looks like this:
<img width="692" alt="image" src="https://github.com/user-attachments/assets/0c83c5ed-f333-42e3-add2-93bcae4a5502">

javiereguiluz commented:
Looks very nice!! Thanks a lot Hugo.

kbond commented:
This looks great!

@Kocal:
1. Can you revert the temporary commit?
2. The proper permissions are still todo?

Kocal commented:
@kbond yes there are still permission issues, and I can revert the temporary commit (when permissions will be fixed, otherwise we won't see the comment)

Kocal commented:
I've reworked the workflow in two dedicated workflows, hoping for permissions issues to be gone.
We won't have comment before the diff is generated, or if a failure happened.

I reverted the commit that modify dist files for testing purposes. I'm merging ASAP
